symbolic-debuginfo/src/bcsymbolmap.rs

```diff
@@ -0,0 +1,215 @@
+//! The Apple [`BCSymbolMap`] file format.
+//!
+//! The [`BCSymbolMap`] is used in Apple bitcode builds when symbols are obfuscated.  In
+//! this case the actual symbol names are replaced with `__hidden#[0-9]+_` in the binary and
+//! dSYMs.  These can then be looked up in the BCSymbolMap to get the original symbol name
+//! back.
+//!
+//! See [`MachObject::load_symbolmap`](crate::macho::MachObject::load_symbolmap) for an
+//! example of how to use this.
+
+use std::error::Error;
+use std::fmt;
+use std::iter::FusedIterator;
+
+use thiserror::Error;
+
+use symbolic_common::DebugId;
+
+const BC_SYMBOL_MAP_HEADER: &str = "BCSymbolMap Version: 2.0";
+
+/// The error type for handling a [`BCSymbolMap`].
+#[derive(Debug, Error)]
+#[error("{kind}")]
+pub struct BCSymbolMapError {
+    kind: BCSymbolMapErrorKind,
+    #[source]
+    source: Option<Box<dyn Error + Send + Sync + 'static>>,
+}
+
+/// Error kind for [`BCSymbolMapError`].
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum BCSymbolMapErrorKind {
+    /// The BCSymbolMap header does not match a supported version.
+    ///
+    /// It could be entirely missing, or only be an unknown version or otherwise corrupted.
+    InvalidHeader,
+    /// The bitcode symbol map did contain invalid UTF-8.
+    InvalidUtf8,
+}
+
+impl fmt::Display for BCSymbolMapErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::InvalidHeader => write!(f, "no valid BCSymbolMap header was found"),
+            Self::InvalidUtf8 => write!(f, "BCSymbolmap is not valid UTF-8"),
+        }
+    }
+}
+
+/// An in-memory representation of the Apple bitcode symbol map.
+///
+/// This is an auxiliary file, not an object file.  It can be used to provide de-obfuscated
+/// symbol names to a [`MachObject`] object using its
+/// [`load_symbolmap`](crate::macho::MachObject::load_symbolmap) method.
+///
+/// It is common for bitcode builds to obfuscate the names in the object file's symbol table
+/// so that even the DWARF files do not have the actual symbol names.  In this case the
+/// build process will create a `.bcsymbolmap` file which maps the obfuscated symbol names
+/// back to the original ones.  This structure can parse these files and allows providing
+/// this information to the [`MachObject`] so that it has the original symbol names instead
+/// of `__hidden#NNN_` ones.
+///
+/// [`MachObject`]: crate::macho::MachObject
+#[derive(Clone, Debug)]
+pub struct BCSymbolMap<'d> {
+    id: DebugId, // TODO
```

Comment:
decide this

Reply:
It looks like there's not really a benefit to having it, so we can as well get rid of it here.

Reply:
`ObjectLike::debug_id()` identifies the object.  It is nice to be able to identify a thing, if not someone who has this need has to create their own wrapper, which admittedly is trivial.  It turned out not to be any burden to require this when creating the `BCSymbolMap`, and one could even imagine a constructor like `from_path()` (though because some users might want async and some not, it's probably not so useful).

Anyway, I can't decide which one I prefer and you seem to think without so let's do without.